src/transformers/models/internlm3/modular_internlm3.py

```diff
@@ -0,0 +1,311 @@
+    
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[Cache] = None,
+        output_attentions: Optional[bool] = False,
+        use_cache: Optional[bool] = False,
+        cache_position: Optional[torch.LongTensor] = None,
+        position_embeddings: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,  # necessary, but kept here for BC
+        **kwargs: Unpack[FlashAttentionKwargs],
+    ) -> Tuple[torch.FloatTensor, Optional[Tuple[torch.FloatTensor, torch.FloatTensor]]]:
+        assert self.hidden_factor == self.config.hidden_factor, (
+            f"The `hidden_factor` in the loaded checkpoint is {self.hidden_factor.item()}, "
+            f"but the `hidden_factor` in the config is {self.config.hidden_factor}."
+        )
+
+        residual = hidden_states
+
+        hidden_states = self.input_layernorm(hidden_states)
+
+        # Self Attention
+        hidden_states, self_attn_weights = self.self_attn(
+            hidden_states=hidden_states,
+            attention_mask=attention_mask,
+            position_ids=position_ids,
+            past_key_value=past_key_value,
+            output_attentions=output_attentions,
+            use_cache=use_cache,
+            cache_position=cache_position,
+            position_embeddings=position_embeddings,
+            **kwargs,
+        )
+        hidden_states = residual + hidden_states
+
+        # Fully Connected
+        residual = hidden_states
+        hidden_states = self.post_attention_layernorm(hidden_states)
+        hidden_states = self.mlp(hidden_states)
+        # scaling hidden states after ffn
+        hidden_states = residual + hidden_states * self.hidden_factor
+
+        outputs = (hidden_states,)
+        if output_attentions:
+            outputs += (self_attn_weights,)
+
+        return outputs
+
+
+class InternLM3PreTrainedModel(LlamaPreTrainedModel):
+    def _init_weights(self, module):
+        std = self.config.initializer_range
+        init_scale_factor = self.config.init_scale_factor
+        if isinstance(module, nn.Linear):
+            module.weight.data.normal_(mean=0.0, std=std)
+            if any(x in module.name for x in ("v_proj", "o_proj", "gate_proj", "up_proj", "down_proj")):
+                module.weight.data.div_(init_scale_factor)
+            if module.bias is not None:
+                module.bias.data.zero_()
+                # 0 / init_scale_factor = 0
+        elif isinstance(module, nn.Embedding):
+            module.weight.data.normal_(mean=0.0, std=std)
+            if module.padding_idx is not None:
+                module.weight.data[module.padding_idx].zero_()
+
+
+class InternLM3Model(LlamaModel):
+    def __init__(self, config: InternLM3Config):
+        super().__init__()
+        # Initialize weights and apply final processing
```

Comment:
not sure I understand what's the final processing? but AFAIK this should be removed as you should not need that. When you check the module itself you can check rather the parent layer. Instead of checking if nn.Linear you check if InternLMAttention 😉